File: src/php/BenchmarkBase.php
<?php

namespace EntityFX\NetBenchmark\Core\Generic {

    use EntityFX\NetBenchmark\Core\Writer;

if (!defined("ASPECT_RATIO")) {
    define("ASPECT_RATIO", 0.2);
}

if (!defined("DEBUG_ASPECT_RATIO")) {
    define("DEBUG_ASPECT_RATIO", 0.05);
}


if (defined("DEBUG")) {
    if (DEBUG) {
        define("BENCH_ASPECT_RATIO", DEBUG_ASPECT_RATIO);
    } else {
        define("BENCH_ASPECT_RATIO", ASPECT_RATIO);
    }
} else {
    define("BENCH_ASPECT_RATIO", ASPECT_RATIO);
}


    
    abstract class BenchmarkBase {
        protected $Iterrations = 0;

        protected $printToConsole = true;

        public static $AspectRatio = BENCH_ASPECT_RATIO;
    
        public $Ratio = 1.0;
    
        public $Name = "";

        protected $output;

        public function __construct($writer, $printToConsole) {
            $this->printToConsole = ($printToConsole === null) ? true : $printToConsole;
            $this->Iterrations = 1;
            $this->Radio = 1.0;
            $this->Name = (new \ReflectionClass($this))->getShortName();
            $this->output = $writer;
        }

        public function Bench() {
            $this->BeforeBench();
            $start = microtime(true);
            $res = $this->BenchImplementation();
            $elapsed = microtime(true) - $start;
            $result = $this->PopulateResult($this->BuildResult($elapsed), $res);
            if ($result["Output"]) {
                file_put_contents("$this->Name.log", $result["Output"], FILE_APPEND);
            }
            $this->AfterBench($result);
            return $result;
        }

        public function BenchInParallel($buildFunc, $benchFunc, $setBenchResultFunc) {
            $results = [];

            return $results;
        }

        protected function BeforeBench() {
        }

        protected function AfterBench() {
        }

        public function Warmup($aspect = 0.05) {
            $aspect = $aspect != null ? $aspect : 0.05;

            $this->Iterrations *= self::$AspectRatio;

            $tmp = $this->Iterrations;
            $this->Iterrations = $this->Iterrations * $aspect;
            $this->UseConsole(DEBUG);
            $this->Bench();
            $this->UseConsole(true);
            $this->Iterrations = $tmp;
        }
        
        public function UseConsole($printToConsole) {
            $this->printToConsole = $printToConsole != null ? $printToConsole : true;
            $this->output->UseConsole = $printToConsole;
        }
        
        protected function BenchImplementation() {
        }

        public static function getCores() {
            return (int) ((substr(PHP_OS, 0, 3) === 'WIN')?(getenv("NUMBER_OF_PROCESSORS")+0):substr_count(file_get_contents("/proc/cpuinfo"),"processor"));
        }
        
        
        protected function BuildResult($elapsed)
        {
            return [
                "BenchmarkName" => $this->Name, 
                "Elapsed" => $elapsed * 1000,
                "Points" => $this->Iterrations / ($elapsed * 1000) * $this->Ratio,
                "Result" => $this->Iterrations / ($elapsed * 1000),
                "Units" => "Iter/s",
                "Iterrations" => $this->Iterrations,
                "Ratio" => $this->Ratio,
                "Output" => ""
            ];
        }
        
        protected function PopulateResult($benchResult, $benchValue)
        {
            return $benchResult;
        }
    }
}

File: src/lua/benchmarkBase.lua
BenchmarkBase = class(function(b, writer, printToConsole)
    b.iterrations = 0
    b.printToConsole = true
    b.isParallel = false
    b.ratio = 1.0
    b.name = "BenchmarkBase"
    b.output = writer
    b.clockBefore = 0
    b.clockAfter = 0
    b.clockDiff = 0
    b.elapsed = 0
end)

BenchmarkBase.IterrationsRatio = 1;

function BenchmarkBase:bench()
    self.clockBefore = 0
    self.clockAfter = 0
    self.clockDiff = 0
    self:beforeBench()
    self.clockBefore = clock()
    local res = self:benchImplementation()
    self.clockAfter = clock()
    self.elapsed = math.floor(self.clockAfter - self.clockBefore)

    local result = self:populateResult(self:buildResult(start), res)
    self:doOutput(result)
    self:afterBench(result)
    return result
end

function BenchmarkBase:warmup(aspect)
    self.iterrations = self.iterrations * BenchmarkBase.IterrationsRatio
    tmp = self.iterrations
    self.iterrations = math.floor(self.iterrations * aspect)
    self:useConsole(false)
    self:bench()
    self:useConsole(true)
    self.iterrations = tmp
end

function BenchmarkBase:beforeBench()
end

function BenchmarkBase:afterBench(result)
end

function BenchmarkBase:useConsole(printToConsole)
    self.printToConsole = printToConsole
    self.output.useConsole = printToConsole
end

function BenchmarkBase:doOutput(result)
    if not result.output then
        return
    end
    stream = io.open(self.name .. ".log", "a")
    stream:write(result.output)
    stream:close()
end

function BenchmarkBase:benchImplementation()
end

function BenchmarkBase:buildResult(start)
    local tElapsed = 0

    if self.elapsed == 0 then
        tElapsed = 1000
    else
        tElapsed = self.elapsed
    end

    local elapsedSeconds = tElapsed / 1000

    return {
        benchmarkName = self.name,
        elapsed = tElapsed,
        points = self.iterrations / tElapsed * self.ratio,
        result = self.iterrations / elapsedSeconds,
        units = "Iter/s",
        iterrations = self.iterrations,
        ratio = self.ratio
    }
end

function BenchmarkBase:populateResult(benchResult, result)
    if isArray(result) then
    end
    return benchResult
end
